File: Client/PL3_en/frameworks/cocos2d-x/cocos/scripting/lua-bindings/auto/api/MatrixView.lua

--------------------------------
-- @module MatrixView
-- @extend ScrollView
-- @parent_module mc

--------------------------------
-- 
-- @function [parent=#MatrixView] doLayout 
-- @param self
-- @return MatrixView#MatrixView self (return value: MatrixView)
        
--------------------------------
-- 
-- @function [parent=#MatrixView] setRowCol 
-- @param self
-- @param #int rowCount
-- @param #int colCount
-- @return MatrixView#MatrixView self (return value: MatrixView)
        
--------------------------------
-- 
-- @function [parent=#MatrixView] removeCustomItem 
-- @param self
-- @param #cc.Node item
-- @return MatrixView#MatrixView self (return value: MatrixView)
        
--------------------------------
-- 
-- @function [parent=#MatrixView] pushBackCustomItem 
-- @param self
-- @param #ccui.Widget item
-- @return MatrixView#MatrixView self (return value: MatrixView)
        
--------------------------------
-- 
-- @function [parent=#MatrixView] create 
-- @param self
-- @param #int rowCount
-- @param #int colCount
-- @return MatrixView#MatrixView ret (return value: MatrixView)
        
--------------------------------
-- @overload self, cc.Node, int, string         
-- @overload self, cc.Node         
-- @function [parent=#MatrixView] addChild
-- @param self
-- @param #cc.Node child
-- @param #int zOrder
-- @param #string name
-- @return MatrixView#MatrixView self (return value: MatrixView)

--------------------------------
-- 
-- @function [parent=#MatrixView] init 
-- @param self
-- @return bool#bool ret (return value: bool)
        
--------------------------------
-- 
-- @function [parent=#MatrixView] requestDoLayout 
-- @param self
-- @return MatrixView#MatrixView self (return value: MatrixView)
        
--------------------------------
-- 
-- @function [parent=#MatrixView] MatrixView 
-- @param self
-- @param #int rowCount
-- @param #int colCount
-- @return MatrixView#MatrixView self (return value: MatrixView)
        
return nil

File: Client/PL3_client/frameworks/runtime-src/Classes/utilities/MatrixView.cpp
//
//  MatrixView.cpp
//  StarClient
//
//  Created by hankai on 16/4/28.
//
//

#include "MatrixView.h"

MatrixView::MatrixView(int rowCount,int colCount)
:_innerContainerDoLayoutDirty(false)
,_itemsMargin(2){
    _rowCount = rowCount;
    _colCount = colCount;
}

MatrixView* MatrixView::create(int rowCount,int colCount){
    MatrixView* widget = new (std::nothrow) MatrixView(rowCount,colCount);
    if (widget && widget->init())
    {
        widget->autorelease();
        return widget;
    }
    CC_SAFE_DELETE(widget);
    return nullptr;
}

bool MatrixView::init()
{
    if (ScrollView::init())
    {
        setDirection(Direction::VERTICAL);
        return true;
    }
    return false;
}

void MatrixView::setRowCol(int rowCount,int colCount){
    _rowCount = rowCount;
    _colCount = colCount;
    requestDoLayout();
}

void MatrixView::onItemListChanged(){
    _outOfBoundaryAmountDirty = true;
}

void MatrixView::doLayout(){
    
    if(!_innerContainerDoLayoutDirty){
        return;
    }
    
    updateItemsPosition();
    
    _innerContainer->forceDoLayout();
    _innerContainerDoLayoutDirty = false;
}
void MatrixView::requestDoLayout(){
    _innerContainerDoLayoutDirty = true;
}

void MatrixView::addChild(Node* child){
    MatrixView::addChild(child, child->getLocalZOrder(), child->getName());
}

void MatrixView::addChild(Node* child, int zOrder, const std::string &name)
{
    ScrollView::addChild(child, zOrder, name);
    
    Widget* widget = dynamic_cast<Widget*>(child);
    if (nullptr != widget)
    {
        _items.pushBack(widget);
        onItemListChanged();
    }
}

void MatrixView::updateItemsPosition(){
    updateInnerContainerSize();
    
    auto containerSize = getInnerContainerSize();
    auto itemSize = _items.front()->getContentSize();
    
    
    
    
    for (int i = 0; i<_items.size(); ++i) {
        int row = i / _colCount;
        int col = i % _colCount;
        
        auto x = _itemsMargin * (col + 1) + itemSize.width/2 + itemSize.width * col;
        auto y = containerSize.height - itemSize.height - itemSize.height * row - _itemsMargin * (row + 1);
        _items.at(i)->setPosition(Vec2(x,y));
    }
}

void MatrixView::updateInnerContainerSize(){
    if (_items.empty()) {
        return;
    }
    auto count = _items.size();
    int rowCount = _rowCount;
    int colCount = _colCount;
    
    if (colCount == 0) {
        rowCount = 1;
        colCount = _items.size();
    }else if(rowCount == 0){
        rowCount = count / colCount;
        if(count % colCount > 0){
            ++rowCount;
        }
    }
    
    
    //auto defaultSize = getInnerContainerSize();
    auto defaultSize = getContentSize();
    
    auto itemSize = _items.front()->getContentSize();
    
    auto size = Size((colCount) * itemSize.width + (colCount+1) * _itemsMargin, (rowCount) * itemSize.height + (rowCount+1) * _itemsMargin);
    
    if (size.width < defaultSize.width) {
        size.width = defaultSize.width;
    }
    if (size.height < defaultSize.height) {
        size.height = defaultSize.height;
    }
    
    
    setInnerContainerSize(size);
}

void MatrixView::pushBackCustomItem(Widget* item)
{
    addChild(item);
    requestDoLayout();
}

void MatrixView::removeCustomItem(Node* item){
    if(item == nullptr){
        return;
    }
    for (int i = 0; i < _items.size(); ++i) {
        if (item == _items.at(i)) {
            _items.at(i)->removeFromParent();
            _items.erase(i);
            requestDoLayout();
            break;
        }
    }
}